assets/model_monitoring/components/src/shared_utilities/span_tree_utils.py

```diff
@@ -69,26 +78,156 @@ def attributes(self) -> str:
-        """Get the span's input from the attributes field as json string."""
+        """Get the span's input from the events payload field as json string."""
+        if self.events is None or len(self.events) == 0:
+            return self.input_from_attributes()
+        else:
+            events_array: list = json.loads(self.events)
+            if len(events_array) == 0:
+                return self.input_from_attributes()
+            input_event = list(filter(lambda event: event.get("name", None) == APP_TRACES_EVENT_LOG_INPUT_KEY,
+                                      events_array))
+            if input_event is not None and len(input_event) > 0:
+                if input_event[0].get("attributes", None) is None:
+                    return None
+                return input_event[0].get("attributes", None).get("payload", None)
+            else:
+                return None
+
+    def input_from_attributes(self) -> str:
+        """Get input from attributes."""
         if self.attributes is None:
+
             return None  # type: ignore
         attribute_dict: dict = json.loads(self.attributes)
         if attribute_dict is None:
-            return None  # type: ignore
+            return None
         return attribute_dict.get("inputs", None)
 
     @property
     def output(self) -> str:
-        """Get the span's output from the attributes field as json string."""
+        """Get the span's output from the events payload field as json string."""
+        if self.events is None or len(self.events) == 0:
+            return self.output_from_attribute()
+        else:
+            events_array: list = json.loads(self.events)
+            if len(events_array) == 0:
+                return self.output_from_attribute()
+            output_event = list(filter(lambda event: event.get("name", None) == APP_TRACES_EVENT_LOG_OUTPUT_KEY,
+                                       events_array))
+            if output_event is not None and len(output_event) > 0:
+                if output_event[0].get("attributes", None) is None:
+                    return None
+                return output_event[0].get("attributes", None).get("payload", None)
+            else:
+                return None
+
+    def output_from_attribute(self) -> str:
+        """Get output from attributes."""
         if self.attributes is None:
             return None  # type: ignore
         attribute_dict: dict = json.loads(self.attributes)
         if attribute_dict is None:
-            return None  # type: ignore
+            return None
         return attribute_dict.get("output", None)
 
+    @property
+    def retrieval_query(self) -> str:
+        """Get promptflow.retrieval.query for retrieval span."""
+        if self.span_type != RETRIEVAL_SPAN_TYPE:
+            return None
+        if self.events is None or len(self.events) == 0:
+            return self.retrieval_query_from_attributes()
+        events_array: list = json.loads(self.events)
```

Comment:
should not need json.loads()

Reply:
it is a jsonstring, that we need to loads.